source/extensions/transport_sockets/alts/alts_proxy.h

```diff
@@ -0,0 +1,68 @@
+#pragma once
+
+#include <memory>
+
+#include "absl/status/statusor.h"
+#include "absl/types/span.h"
+#include "grpcpp/channel.h"
+#include "grpcpp/client_context.h"
+#include "grpcpp/support/sync_stream.h"
+#include "src/proto/grpc/gcp/handshaker.grpc.pb.h"
+#include "src/proto/grpc/gcp/handshaker.pb.h"
+#include "src/proto/grpc/gcp/transport_security_common.pb.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace TransportSockets {
+namespace Alts {
+
+constexpr char ApplicationProtocol[] = "grpc";
+constexpr char RecordProtocol[] = "ALTSRP_GCM_AES128_REKEY";
+constexpr std::size_t MaxFrameSize = 1024 * 1024;
+constexpr std::size_t MaxMajorRpcVersion = 2;
+constexpr std::size_t MaxMinorRpcVersion = 1;
+constexpr std::size_t MinMajorRpcVersion = 2;
+constexpr std::size_t MinMinorRpcVersion = 1;
+
+// Manages a bidirectional stream to the ALTS handshaker service. An AltsProxy
+// instance is tied to a single ALTS handshake and must not be reused.
+class AltsProxy {
+public:
+  static absl::StatusOr<std::unique_ptr<AltsProxy>>
+  create(std::shared_ptr<grpc::Channel> handshaker_service_channel);
+
+  ~AltsProxy();
+
+  // Sends a StartClientHandshakeReq message to the ALTS handshaker service and
+  // returns the response. This API is blocking.
+  absl::StatusOr<grpc::gcp::HandshakerResp> sendStartClientHandshakeReq();
+
+  // Sends a StartServerHandshakeReq message to the ALTS handshaker service and
+  // returns the response. This API is blocking.
+  absl::StatusOr<grpc::gcp::HandshakerResp>
+  sendStartServerHandshakeReq(absl::Span<const uint8_t> in_bytes);
+
+  // Sends a NextHandshakeMessageReq message to the ALTS handshaker service and
+  // returns the response. This API is blocking.
```

Comment:
What is being blocked here? e.g. are we blocking the Envoy silo here with any / all of the calls above? If so we can easily bring the system to a halt by the calls above.

Reply:
No sorry, I just mean that we're blocking the current thread. I removed this comment because I don't think it adds anything.

Reply:
@botengyao to correct me if I explain the TSI socket poorly. :)

When the ALTS transport socket is handling an ALTS handshake, this is occurring in a worker thread, and we are block this worker thread when we call this function and wait for this function to execute. Under the hood of this function, we are making a gRPC call to the ATLS handshaker service.

@KBaichoo Does that help clarify? Can I expand on anything?

Apologies I'm not sure what you mean by "the Envoy silo" above. If this does not answer your question, can you expand on this?

Reply:
Thanks for clarifying, as such as can easily bring the entire proxy to a halt if the grpc service is unavailable / experiencing high latency. I think we should strongly document this, and possibly follow up. I suspect mostly Google might use this extension @yanavlasov @adisuissa @botengyao  on if/when this bottleneck would need to be addressed.

Reply:
Right, we should document this. I think the best solution is to make this RPC call as an async stream with response callback, similar as the async gRPC client here. https://github.com/envoyproxy/envoy/blob/1140c93e207bc30995dfb5948bff9e4b48302f6f/source/common/grpc/async_client_impl.h#L12. Did we perform the load test with the current implementation internally with with currently logic?

Reply:
Thansk @KBaichoo for explaining, really glad you caught this. I've added documentation to alts_proxy.h and added a TODO in alts_proxy.cc to further investigate the async client.

@botengyao I have load testing set up with the internal implementation and I did not notice this issue there. For that reason, I'm tempted to suggest that we merge this PR (so that there is a working version of the ALTS transport socket) and look into a follow-up PR to use the async client. WDYT?

Reply:
sgtm, defer to @KBaichoo for approval. Thanks!

Reply:
Having saying that, and taking a deeper look, I think this still should be the **top priority TODO** to make the gRPC call be async to not block the worker thread based on Envoy's silo. I tend to let this PR merged since there is no actual usage and this makes the extension workable at least, but also glad to hear @KBaichoo and others' opinion.

Reply:
+1 to letting this through since I expect there might not be significant usage of this (given the preexisting bug); this is a step in the right direction and this PR is already large enough.